In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from pandas import DataFrame, Series

In [2]:
data = pd.read_csv('data_for_test.csv')

In [3]:
data.head(5)

pipe_id  install_year material  diameter   env_dat_0  env_dat_1  \
0  P0000000000          1985        D     20.32  623.542175  18.052843   
1  P0000000001          1934        C     15.24  644.830017  17.790001   
2  P0000000002          1904        C     20.32  604.570007  18.219999   
3  P0000000003          1979        D     15.24  573.424011  17.671381   
4  P0000000004          1987        D     20.32  583.429993  18.280001   

   env_dat_2  env_dat_3   env_dat_4  env_dat_5  ...  env_dat_136  env_dat_137  \
0  14.161422  10.268579  159.662827   0.010340  ...    20.000000     0.000038   
1  13.830000   9.870000  257.001221   0.007786  ...     9.200573     0.304329   
2  14.350000  10.470000  225.165039   0.014548  ...     4.872639     0.182168   
3  14.132636  10.586646  170.244339   0.014938  ...    12.353504     0.480517   
4  14.350000  10.420000  171.034073   0.025906  ...    25.052542     0.021532   

   env_dat_138  env_dat_139  env_dat_140  env_dat_141  env_dat_142  \
0   438.628143    27.000000  3085.786865     0.002728     0.311803   
1   239.719299    32.571018  3188.292236     0.003291     0.322161   
2   500.638367    28.978306  4542.257324     0.002928     0.458972   
3   295.554504    41.983772  4521.720215     0.004242     0.456897   
4   778.687256    25.842377  4978.718262     0.002611     0.503074   

   env_dat_143  env_dat_144  target  
0          0.0          0.0       0  
1          0.0          0.0       0  
2          0.0          0.0       0  
3          0.0          0.0       0  
4          0.0          0.0       0  

[5 rows x 150 columns]

In [4]:
y = data["target"]
X = data.drop("target", axis=1)

In [5]:
y.unique()

array([0, 1])

In [6]:
X.shape

(56774, 149)

In [7]:
X.drop_duplicates(keep = 'last')

pipe_id  install_year material  diameter   env_dat_0  env_dat_1  \
0      P0000000000          1985        D     20.32  623.542175  18.052843   
1      P0000000001          1934        C     15.24  644.830017  17.790001   
2      P0000000002          1904        C     20.32  604.570007  18.219999   
3      P0000000003          1979        D     15.24  573.424011  17.671381   
4      P0000000004          1987        D     20.32  583.429993  18.280001   
...            ...           ...      ...       ...         ...        ...   
56769  P0000055554          2016        D     20.32  610.280029  18.280001   
56770  P0000055555          2016        D     20.32  610.280029  18.280001   
56771  P0000055556          2016        D     20.32  610.280029  18.280001   
56772  P0000055557          2016        D     20.32  610.280029  18.280001   
56773  P0000055558          2016        D     20.32  610.280029  18.280001   

       env_dat_2  env_dat_3   env_dat_4  env_dat_5  ...  env_dat_135  \
0      14.161422  10.268579  159.662827   0.010340  ...   554.479675   
1      13.830000   9.870000  257.001221   0.007786  ...   276.357605   
2      14.350000  10.470000  225.165039   0.014548  ...   236.882950   
3      14.132636  10.586646  170.244339   0.014938  ...   321.886658   
4      14.350000  10.420000  171.034073   0.025906  ...   758.841431   
...          ...        ...         ...        ...  ...          ...   
56769  14.340000  10.400000  190.049469   0.010628  ...   396.748718   
56770  14.340000  10.400000  190.049469   0.010628  ...   396.748718   
56771  14.340000  10.400000  190.049469   0.010628  ...   396.748718   
56772  14.340000  10.400000  190.049469   0.010628  ...   396.748718   
56773  14.340000  10.400000  190.049469   0.010628  ...   396.748718   

       env_dat_136  env_dat_137  env_dat_138  env_dat_139  env_dat_140  \
0        20.000000     0.000038   438.628143    27.000000  3085.786865   
1         9.200573     0.304329   239.719299    32.571018  3188.292236   
2         4.872639     0.182168   500.638367    28.978306  4542.257324   
3        12.353504     0.480517   295.554504    41.983772  4521.720215   
4        25.052542     0.021532   778.687256    25.842377  4978.718262   
...            ...          ...          ...          ...          ...   
56769     6.000000     0.719997   170.973465    26.000000  2504.886719   
56770     6.000000     0.719997   170.973465    26.000000  2504.886719   
56771     6.000000     0.719997   170.973465    26.000000  2504.886719   
56772     6.000000     0.719997   170.973465    26.000000  2504.886719   
56773     6.000000     0.719997   170.973465    26.000000  2504.886719   

       env_dat_141  env_dat_142  env_dat_143  env_dat_144  
0         0.002728     0.311803          0.0          0.0  
1         0.003291     0.322161          0.0          0.0  
2         0.002928     0.458972          0.0          0.0  
3         0.004242     0.456897          0.0          0.0  
4         0.002611     0.503074          0.0          0.0  
...            ...          ...          ...          ...  
56769     0.002627     0.253106          0.0          0.0  
56770     0.002627     0.253106          0.0          0.0  
56771     0.002627     0.253106          0.0          0.0  
56772     0.002627     0.253106          0.0          0.0  
56773     0.002627     0.253106          0.0          0.0  

[55559 rows x 149 columns]

In [8]:
X[X.isna().any(axis=1)]

pipe_id  install_year material  diameter   env_dat_0  env_dat_1  \
9      P0000000009          1967        C     40.64  621.520020  18.389999   
121    P0000000119          1934        C       NaN  635.059998  17.490000   
454    P0000000452          1989        D       NaN  615.719971  17.959999   
519    P0000000515          1933        C       NaN  632.330017  18.520000   
672    P0000000666          1934        C       NaN  635.059998  17.490000   
...            ...           ...      ...       ...         ...        ...   
56366  P0000055151          2002        D     30.48  552.996704  17.968988   
56558  P0000055343          2016        D     30.48  538.739990  18.580000   
56559  P0000055344          2016        D     20.32  538.739990  18.580000   
56560  P0000055345          2016        D     20.32  538.739990  18.580000   
56561  P0000055346          2016        D     30.48  538.739990  18.580000   

       env_dat_2  env_dat_3   env_dat_4  env_dat_5  ...  env_dat_135  \
9      14.380000  10.370000  122.000000   0.002382  ...          NaN   
121    13.850000  10.210000  213.618225   0.010895  ...   202.541885   
454    14.090000  10.210000   34.133255   0.002201  ...   143.234604   
519    14.410000  10.290000  204.133987   0.011457  ...   299.100677   
672    13.850000  10.210000   89.535225   0.004574  ...   344.991058   
...          ...        ...         ...        ...  ...          ...   
56366  14.361272  10.750011   11.477139   0.000139  ...   854.363037   
56558  14.700000  10.810000  243.621994   0.002717  ...          NaN   
56559  14.700000  10.810000  243.621994   0.002717  ...          NaN   
56560  14.700000  10.810000  243.621994   0.002717  ...          NaN   
56561  14.700000  10.810000  243.621994   0.002717  ...          NaN   

       env_dat_136  env_dat_137  env_dat_138  env_dat_139  env_dat_140  \
9              NaN     8.491898   120.759766    15.000000   835.561951   
121       2.000000     0.092864   153.147385    41.000000  3679.603027   
454       6.000000     5.374597   463.910614    13.000000  2938.658936   
519       9.000000     1.401415   239.565048    31.000000  2849.369141   
672       9.956148     3.996830  1200.711426    19.361086  2969.813477   
...            ...          ...          ...          ...          ...   
56366    29.000000          NaN          NaN          NaN          NaN   
56558          NaN          NaN          NaN          NaN          NaN   
56559          NaN          NaN          NaN          NaN          NaN   
56560          NaN          NaN          NaN          NaN          NaN   
56561          NaN          NaN          NaN          NaN          NaN   

       env_dat_141  env_dat_142  env_dat_143  env_dat_144  
9         0.001516     0.084429          NaN          NaN  
121       0.004143     0.371805          0.0          0.0  
454       0.001314     0.296937          0.0          0.0  
519       0.003132     0.287914          0.0          0.0  
672       0.001956     0.300085          0.0          0.0  
...            ...          ...          ...          ...  
56366          NaN          NaN          0.0          0.0  
56558          NaN          NaN          NaN          NaN  
56559          NaN          NaN          NaN          NaN  
56560          NaN          NaN          NaN          NaN  
56561          NaN          NaN          NaN          NaN  

[824 rows x 149 columns]

In [9]:
X[X['material'].isna()]

Empty DataFrame
Columns: [pipe_id, install_year, material, diameter, env_dat_0, env_dat_1, env_dat_2, env_dat_3, env_dat_4, env_dat_5, env_dat_6, env_dat_7, env_dat_8, env_dat_9, env_dat_10, env_dat_11, env_dat_12, env_dat_13, env_dat_14, env_dat_15, env_dat_16, env_dat_17, env_dat_18, env_dat_19, env_dat_20, env_dat_21, env_dat_22, env_dat_23, env_dat_24, env_dat_25, env_dat_26, env_dat_27, env_dat_28, env_dat_29, env_dat_30, env_dat_31, env_dat_32, env_dat_33, env_dat_34, env_dat_35, env_dat_36, env_dat_37, env_dat_38, env_dat_39, env_dat_40, env_dat_41, env_dat_42, env_dat_43, env_dat_44, env_dat_45, env_dat_46, env_dat_47, env_dat_48, env_dat_49, env_dat_50, env_dat_51, env_dat_52, env_dat_53, env_dat_54, env_dat_55, env_dat_56, env_dat_57, env_dat_58, env_dat_59, env_dat_60, env_dat_61, env_dat_62, env_dat_63, env_dat_64, env_dat_65, env_dat_66, env_dat_67, env_dat_68, env_dat_69, env_dat_70, env_dat_71, env_dat_72, env_dat_73, env_dat_74, env_dat_75, env_dat_76, env_dat_77, env_dat_78, env_dat_79, env_dat_80, env_dat_81, env_dat_82, env_dat_83, env_dat_84, env_dat_85, env_dat_86, env_dat_87, env_dat_88, env_dat_89, env_dat_90, env_dat_91, env_dat_92, env_dat_93, env_dat_94, env_dat_95, ...]
Index: []

[0 rows x 149 columns]

In [10]:
X[X['install_year'].isna()]

Empty DataFrame
Columns: [pipe_id, install_year, material, diameter, env_dat_0, env_dat_1, env_dat_2, env_dat_3, env_dat_4, env_dat_5, env_dat_6, env_dat_7, env_dat_8, env_dat_9, env_dat_10, env_dat_11, env_dat_12, env_dat_13, env_dat_14, env_dat_15, env_dat_16, env_dat_17, env_dat_18, env_dat_19, env_dat_20, env_dat_21, env_dat_22, env_dat_23, env_dat_24, env_dat_25, env_dat_26, env_dat_27, env_dat_28, env_dat_29, env_dat_30, env_dat_31, env_dat_32, env_dat_33, env_dat_34, env_dat_35, env_dat_36, env_dat_37, env_dat_38, env_dat_39, env_dat_40, env_dat_41, env_dat_42, env_dat_43, env_dat_44, env_dat_45, env_dat_46, env_dat_47, env_dat_48, env_dat_49, env_dat_50, env_dat_51, env_dat_52, env_dat_53, env_dat_54, env_dat_55, env_dat_56, env_dat_57, env_dat_58, env_dat_59, env_dat_60, env_dat_61, env_dat_62, env_dat_63, env_dat_64, env_dat_65, env_dat_66, env_dat_67, env_dat_68, env_dat_69, env_dat_70, env_dat_71, env_dat_72, env_dat_73, env_dat_74, env_dat_75, env_dat_76, env_dat_77, env_dat_78, env_dat_79, env_dat_80, env_dat_81, env_dat_82, env_dat_83, env_dat_84, env_dat_85, env_dat_86, env_dat_87, env_dat_88, env_dat_89, env_dat_90, env_dat_91, env_dat_92, env_dat_93, env_dat_94, env_dat_95, ...]
Index: []

[0 rows x 149 columns]

In [11]:
X.iloc[0,4:]

env_dat_0       623.542175
env_dat_1        18.052843
env_dat_2        14.161422
env_dat_3        10.268579
env_dat_4       159.662827
                  ...     
env_dat_140    3085.786865
env_dat_141       0.002728
env_dat_142       0.311803
env_dat_143            0.0
env_dat_144            0.0
Name: 0, Length: 145, dtype: object

In [12]:
X.iloc[:,0:4]

pipe_id  install_year material  diameter
0      P0000000000          1985        D     20.32
1      P0000000001          1934        C     15.24
2      P0000000002          1904        C     20.32
3      P0000000003          1979        D     15.24
4      P0000000004          1987        D     20.32
...            ...           ...      ...       ...
56769  P0000055554          2016        D     20.32
56770  P0000055555          2016        D     20.32
56771  P0000055556          2016        D     20.32
56772  P0000055557          2016        D     20.32
56773  P0000055558          2016        D     20.32

[56774 rows x 4 columns]

In [13]:
from sklearn.preprocessing import OrdinalEncoder
ordinal = OrdinalEncoder()
X_categorical = ordinal.fit_transform(X.iloc[:,0:3])

In [14]:
from sklearn.impute import SimpleImputer
imp = SimpleImputer(strategy="mean")
X_numerical = imp.fit_transform(X.iloc[:,4:])

In [15]:
X_all = np.concatenate((X_categorical, X_numerical), axis=1)

In [16]:
X_all.shape

(56774, 148)

In [17]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_all, y, test_size=0.2, random_state=42)

In [18]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix, hamming_loss

ModelsPerformance = {}
def metricsReport(modelName, test_labels, predictions):
    accuracy = accuracy_score(test_labels, predictions)

    macro_precision = precision_score(test_labels, predictions, average='macro')
    macro_recall = recall_score(test_labels, predictions, average='macro')
    macro_f1 = f1_score(test_labels, predictions, average='macro')

    micro_precision = precision_score(test_labels, predictions, average='micro')
    micro_recall = recall_score(test_labels, predictions, average='micro')
    micro_f1 = f1_score(test_labels, predictions, average='micro')
#     hamLoss = hamming_loss(test_labels, predictions)
    print("------" + modelName + " Model Metrics-----")
    print("Accuracy: {:.4f}\nPrecision:\n  - Macro: {:.4f}\n  - Micro: {:.4f}\nRecall:\n  - Macro: {:.4f}\n  - Micro: {:.4f}\nF1-measure:\n  - Macro: {:.4f}\n  - Micro: {:.4f}"\
          .format(accuracy, macro_precision, micro_precision, macro_recall, micro_recall, macro_f1, micro_f1))
    ModelsPerformance[modelName] = micro_f1

## Random Forrest

Random forests provide an improvement over bagged trees by way of a small tweak that decorrelates the trees. As in bagging, we build a number of decision trees on bootstrapped training samples. But when building these decision trees, each time a split in a tree is considered, a random sample of m predictors is chosen as split candidates from the full set of p predictors. The split is allowed to use only one of those m predictors. A fresh sample of √m predictors is taken at each split, and typically we choose m ≈ p—that is, the number of predictors considered at each split is approximately equal to the square root of the total number of predictors

In other words, in building a random forest, at each split in the tree, the algorithm is not even allowed to consider a majority of the available predictors.

In [19]:
from sklearn.ensemble import RandomForestClassifier
rfClassifier = RandomForestClassifier(n_jobs=-1)
rfClassifier.fit(X_train, y_train)
rfPreds = rfClassifier.predict(X_test)
metricsReport("Random Forest", y_test, rfPreds)

------Random Forest Model Metrics-----
Accuracy: 0.9383
Precision:
  - Macro: 0.6791
  - Micro: 0.9383
Recall:
  - Macro: 0.6016
  - Micro: 0.9383
F1-measure:
  - Macro: 0.6273
  - Micro: 0.9383


## k-nearest neighbors algorithm (kNN)

is a non-parametric technique used for classification. Given a test document x, the KNN algorithm finds the k nearest neighbors of x among all the documents in the training set, and scores the category candidates based the class of k neighbors

In [20]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.multiclass import OneVsRestClassifier

knnClf = KNeighborsClassifier()

knnClf.fit(X_train, y_train)
knnPredictions = knnClf.predict(X_test)
metricsReport("knn", y_test, knnPredictions)

------knn Model Metrics-----
Accuracy: 0.9373
Precision:
  - Macro: 0.6240
  - Micro: 0.9373
Recall:
  - Macro: 0.5402
  - Micro: 0.9373
F1-measure:
  - Macro: 0.5548
  - Micro: 0.9373


## Decision Tree Classifier

Tree-based methods are simple and useful for interpretation. A decision tree is a flowchart-like structure in which each internal node represents a "test" on an attribute (e.g. whether a coin flip comes up heads or tails), each branch represents the outcome of the test, and each leaf node represents a class label (decision taken after computing all attributes). The paths from root to leaf represent classification rules.

In [21]:
from sklearn.tree import DecisionTreeClassifier

dtClassifier = DecisionTreeClassifier()
dtClassifier.fit(X_train, y_train)
dtPreds = dtClassifier.predict(X_test)
metricsReport("Decision Tree", y_test, dtPreds)

------Decision Tree Model Metrics-----
Accuracy: 0.9114
Precision:
  - Macro: 0.5915
  - Micro: 0.9114
Recall:
  - Macro: 0.5986
  - Micro: 0.9114
F1-measure:
  - Macro: 0.5948
  - Micro: 0.9114


## Bagging

The decision trees suffer from high variance.
This means that if we split the training data into two parts at random, and fit a decision tree to both halves, the results that we get could be quite different.
Bootstrap aggregation, or bagging, is a general-purpose procedure for reducing the variance of a statistical learning method; we introduce it here because it is particularly useful and frequently used in the context of decision trees.

In [22]:
from sklearn.ensemble import BaggingClassifier
y_test
bagClassifier = OneVsRestClassifier(BaggingClassifier(n_jobs=-1))
bagClassifier.fit(X_train, y_train)
bagPreds = bagClassifier.predict(X_test)
metricsReport("Bagging", y_test, bagPreds)

------Bagging Model Metrics-----
Accuracy: 0.9398
Precision:
  - Macro: 0.6946
  - Micro: 0.9398
Recall:
  - Macro: 0.6180
  - Micro: 0.9398
F1-measure:
  - Macro: 0.6450
  - Micro: 0.9398


## Boosting

Recall that bagging involves creating multiple copies of the original train- ing data set using the bootstrap, fitting a separate decision tree to each copy, and then combining all of the trees in order to create a single predic- tive model. Notably, each tree is built on a bootstrap data set, independent of the other trees. Boosting works in a similar way, except that the trees are grown sequentially: each tree is grown using information from previously grown trees. Boosting does not involve bootstrap sampling; instead each tree is fit on a modified version of the original data set.

In [23]:
from sklearn.ensemble import GradientBoostingClassifier

boostClassifier = OneVsRestClassifier(GradientBoostingClassifier())
boostClassifier.fit(X_train, y_train)
boostPreds = boostClassifier.predict(X_test)
metricsReport("Boosting", y_test, boostPreds)

------Boosting Model Metrics-----
Accuracy: 0.9442
Precision:
  - Macro: 0.7194
  - Micro: 0.9442
Recall:
  - Macro: 0.5230
  - Micro: 0.9442
F1-measure:
  - Macro: 0.5301
  - Micro: 0.9442


In [24]:
from sklearn.naive_bayes import GaussianNB

nbClassifier = OneVsRestClassifier(GaussianNB())
nbClassifier.fit(X_train, y_train)

nbPreds = nbClassifier.predict(X_test)
metricsReport("GaussianNB", y_test, nbPreds)

------GaussianNB Model Metrics-----
Accuracy: 0.4881
Precision:
  - Macro: 0.5216
  - Micro: 0.4881
Recall:
  - Macro: 0.6018
  - Micro: 0.4881
F1-measure:
  - Macro: 0.3866
  - Micro: 0.4881


## Support Vector Machine (Linear SVC)

Support Vector Machine (SVM), an approach for classification that was developed in the computer science community in the 1990s and that has grown in popularity since then.
SVMs have been shown to perform well in a variety of settings, and are often considered one of the best “out of the box” classifiers.
The support vector machine is a generalization of a simple and intuitive classifier called the maximal margin classifier

In [25]:
from sklearn.svm import LinearSVC

svmClassifier = OneVsRestClassifier(LinearSVC(), n_jobs=-1)
svmClassifier.fit(X_train, y_train)
svmPreds = svmClassifier.predict(X_test)
metricsReport("SVC Sq. Hinge Loss", y_test, svmPreds)

------SVC Sq. Hinge Loss Model Metrics-----
Accuracy: 0.9442
Precision:
  - Macro: 0.4721
  - Micro: 0.9442
Recall:
  - Macro: 0.5000
  - Micro: 0.9442
F1-measure:
  - Macro: 0.4856
  - Micro: 0.9442


## Comparison on different models based on their Micro-F1 score

In [26]:
print("  Model Name " + " "*10 + "| Micro-F1 Score")
print("-------------------------------------------")
for key, value in ModelsPerformance.items():
    print("  " + key, " "*(20-len(key)) + "|", value)
    print("-------------------------------------------")

  Model Name           | Micro-F1 Score
-------------------------------------------
  Random Forest        | 0.938265081461911
-------------------------------------------
  knn                  | 0.937296345222369
-------------------------------------------
  Decision Tree        | 0.911404667547336
-------------------------------------------
  Bagging              | 0.9397622192866578
-------------------------------------------
  Boosting             | 0.9441655658300309
-------------------------------------------
  GaussianNB           | 0.4880669308674593
-------------------------------------------
  SVC Sq. Hinge Loss   | 0.9441655658300309
-------------------------------------------


In [27]:
data[data["target"]==0].shape

(53630, 150)

## The Data imbalance problem

In [45]:
from collections import Counter
from sklearn.datasets import make_classification
from imblearn.over_sampling import SMOTE
from imblearn.under_sampling import RandomUnderSampler
from imblearn.pipeline import Pipeline
from matplotlib import pyplot
from numpy import where

In [46]:
y = data["target"]
X = data.drop("target", axis=1)
ordinal = OrdinalEncoder()
X_categorical = ordinal.fit_transform(X.iloc[:,0:3])
imp = SimpleImputer(strategy="mean")
X_numerical = imp.fit_transform(X.iloc[:,4:])
X_all = np.concatenate((X_categorical, X_numerical), axis=1)

In [31]:
# summarize class distribution
counter = Counter(y)
print(counter)

Counter({0: 53630, 1: 3144})


In [34]:
# define pipeline
over = SMOTE(sampling_strategy=0.1)
under = RandomUnderSampler(sampling_strategy=0.5)
steps = [('o', over), ('u', under)]
pipeline = Pipeline(steps=steps)
# transform the dataset
X_all, y = pipeline.fit_resample(X_all, y)
# summarize the new class distribution
counter = Counter(y)
print(counter)

Counter({0: 10726, 1: 5363})


In [35]:
# decision tree evaluated on imbalanced dataset
from numpy import mean
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import RepeatedStratifiedKFold
from sklearn.tree import DecisionTreeClassifier
# define model
model = DecisionTreeClassifier()
# evaluate pipeline
cv = RepeatedStratifiedKFold(n_splits=10, n_repeats=3, random_state=1)
scores = cross_val_score(model, X_all, y, scoring='roc_auc', cv=cv, n_jobs=-1)
print('Mean ROC AUC: %.3f' % mean(scores))

Mean ROC AUC: 0.778


In [48]:
y = data["target"]
X = data.drop("target", axis=1)
ordinal = OrdinalEncoder()
X_categorical = ordinal.fit_transform(X.iloc[:,0:3])
imp = SimpleImputer(strategy="mean")
X_numerical = imp.fit_transform(X.iloc[:,4:])
X_all = np.concatenate((X_categorical, X_numerical), axis=1)
# define pipeline
steps = [('over', SMOTE()), ('model', DecisionTreeClassifier())]
pipeline = Pipeline(steps=steps)
# evaluate pipeline
cv = RepeatedStratifiedKFold(n_splits=10, n_repeats=3, random_state=1)
scores = cross_val_score(pipeline, X_all, y, scoring='roc_auc', cv=cv, n_jobs=-1)
print('Mean ROC AUC: %.3f' % mean(scores))

Mean ROC AUC: 0.701


As mentioned in the paper, it is believed that SMOTE performs better when combined with undersampling of the majority class, such as random undersampling.

I achieve this by simply adding a RandomUnderSampler step to the Pipeline.

In [49]:
y = data["target"]
X = data.drop("target", axis=1)
ordinal = OrdinalEncoder()
X_categorical = ordinal.fit_transform(X.iloc[:,0:3])
imp = SimpleImputer(strategy="mean")
X_numerical = imp.fit_transform(X.iloc[:,4:])
X_all = np.concatenate((X_categorical, X_numerical), axis=1)
# define pipeline
model = DecisionTreeClassifier()
over = SMOTE(sampling_strategy=0.1)
under = RandomUnderSampler(sampling_strategy=0.5)
steps = [('over', over), ('under', under), ('model', model)]
pipeline = Pipeline(steps=steps)

In [50]:
cv = RepeatedStratifiedKFold(n_splits=10, n_repeats=3, random_state=1)
scores = cross_val_score(pipeline, X_all, y, scoring='roc_auc', cv=cv, n_jobs=-1)
print('Mean ROC AUC: %.3f' % mean(scores))

Mean ROC AUC: 0.712


I can explore testing different ratios of the minority class and majority class (e.g. changing the sampling_strategy argument) to see if a further lift in performance is possible.

Another area to explore would be to test different values of the k-nearest neighbors selected in the SMOTE procedure when each new synthetic example is created. The default is k=5, although larger or smaller values will influence the types of examples created, and in turn, may impact the performance of the model.

For example, I can grid search a range of values of k, such as values from 1 to 7, and evaluate the pipeline for each value.

Running this will perform SMOTE oversampling with different k values for the KNN used in the procedure, followed by random undersampling and fitting a decision tree on the resulting training dataset.

The mean ROC AUC is reported for each configuration.

In [51]:
y = data["target"]
X = data.drop("target", axis=1)
ordinal = OrdinalEncoder()
X_categorical = ordinal.fit_transform(X.iloc[:,0:3])
imp = SimpleImputer(strategy="mean")
X_numerical = imp.fit_transform(X.iloc[:,4:])
X_all = np.concatenate((X_categorical, X_numerical), axis=1)
# values to evaluate
k_values = [1, 2, 3, 4, 5, 6, 7]
for k in k_values:
    # define pipeline
    model = DecisionTreeClassifier()
    over = SMOTE(sampling_strategy=0.1, k_neighbors=k)
    under = RandomUnderSampler(sampling_strategy=0.5)
    steps = [('over', over), ('under', under), ('model', model)]
    pipeline = Pipeline(steps=steps)
    # evaluate pipeline
    cv = RepeatedStratifiedKFold(n_splits=10, n_repeats=3, random_state=1)
    scores = cross_val_score(pipeline, X_all, y, scoring='roc_auc', cv=cv, n_jobs=-1)
    score = mean(scores)
    print('> k=%d, Mean ROC AUC: %.3f' % (k, score))

> k=1, Mean ROC AUC: 0.710
> k=2, Mean ROC AUC: 0.710
> k=3, Mean ROC AUC: 0.712
> k=4, Mean ROC AUC: 0.712
> k=5, Mean ROC AUC: 0.712
> k=6, Mean ROC AUC: 0.711
> k=7, Mean ROC AUC: 0.713


Note: Your results may vary given the stochastic nature of the algorithm or evaluation procedure, or differences in numerical precision.
Consider running the example a few times and compare the average outcome.
In this case, the results suggest that a k=7 might be good with a ROC AUC of about 0.713

This highlights that both the amount of oversampling and undersampling performed (sampling_strategy argument) and the number of datapoints selected from which a partner is chosen to create a synthetic example (k_neighbors) may be important parameters to select and tune for my dataset.

## SMOTE With Selective Synthetic Sample Generation

I can be selective about the datapoints in the minority class that are oversampled using SMOTE.

In this section, I will review some extensions to SMOTE that are more selective regarding the datapoints from the minority class that provide the basis for generating new synthetic datapoints.

## Borderline-SMOTE
A popular extension to SMOTE involves selecting those instances of the minority class that are misclassified, such as with a k-nearest neighbor classification model.

I can then oversample just those difficult instances, providing more resolution only where it may be required.

>The examples on the borderline and the ones nearby […] are more apt to be misclassified than the ones far from the borderline, and thus more important for classification.
[ Borderline-SMOTE: A New Over-Sampling Method in Imbalanced Data Sets Learning, 2005.]

These datapoints that are misclassified are likely ambiguous and in a region of the edge or border of decision boundary where class membership may overlap.
As such, this modified to SMOTE is called Borderline-SMOTE and was proposed by Hui Han, et al. in their 2005 paper titled “Borderline-SMOTE: A New Over-Sampling Method in Imbalanced Data Sets Learning.”

In [54]:
from imblearn.over_sampling import BorderlineSMOTE

y = data["target"]
X = data.drop("target", axis=1)
ordinal = OrdinalEncoder()
X_categorical = ordinal.fit_transform(X.iloc[:,0:3])
imp = SimpleImputer(strategy="mean")
X_numerical = imp.fit_transform(X.iloc[:,4:])
X_all = np.concatenate((X_categorical, X_numerical), axis=1)

# transform the dataset
oversample = BorderlineSMOTE()
X_all, y = oversample.fit_resample(X_all, y)
# define model and evaluate it
model = DecisionTreeClassifier()
cv = RepeatedStratifiedKFold(n_splits=10, n_repeats=3, random_state=1)
scores = cross_val_score(model, X_all, y, scoring='roc_auc', cv=cv, n_jobs=-1)
print('Mean ROC AUC: %.3f' % mean(scores))

Mean ROC AUC: 0.945


## Borderline-SMOTE SVM
Hien Nguyen, et al. suggest using an alternative of Borderline-SMOTE where an SVM algorithm is used instead of a KNN to identify misclassified datapoints on the decision boundary.

Their approach is summarized in the 2009 paper titled “Borderline Over-sampling For Imbalanced Data Classification.”
An SVM is used to locate the decision boundary defined by the support vectors and datapoints in the minority class that close to the support vectors become the focus for generating synthetic datapoints.

In [55]:
from imblearn.over_sampling import SVMSMOTE

y = data["target"]
X = data.drop("target", axis=1)
ordinal = OrdinalEncoder()
X_categorical = ordinal.fit_transform(X.iloc[:,0:3])
imp = SimpleImputer(strategy="mean")
X_numerical = imp.fit_transform(X.iloc[:,4:])
X_all = np.concatenate((X_categorical, X_numerical), axis=1)

# transform the dataset
oversample = SVMSMOTE()
X_all, y = oversample.fit_resample(X_all, y)

# define model and evaluate it
model = DecisionTreeClassifier()
cv = RepeatedStratifiedKFold(n_splits=10, n_repeats=3, random_state=1)
scores = cross_val_score(model, X_all, y, scoring='roc_auc', cv=cv, n_jobs=-1)
print('Mean ROC AUC: %.3f' % mean(scores))

Mean ROC AUC: 0.934


Using Borderline-SMOTE SVM, unlike Borderline-SMOTE, more datapoints are synthesized away from the region of class overlap, such as toward the top left of the plot.

## Adaptive Synthetic Sampling (ADASYN)
Another approach involves generating synthetic samples inversely proportional to the density of the datapoints in the minority class.

That is, generate more synthetic datapoints in regions of the feature space where the density of minority datapoints is low, and fewer or none where the density is high.

This modification to SMOTE is referred to as the Adaptive Synthetic Sampling Method, or ADASYN, and was proposed to Haibo He, et al. in their 2008 paper named for the method titled “ADASYN: Adaptive Synthetic Sampling Approach For Imbalanced Learning.”

>ADASYN is based on the idea of adaptively generating minority data samples according to their distributions: more synthetic data is generated for minority class samples that are harder to learn compared to those minority samples that are easier to learn.

— ADASYN: Adaptive synthetic sampling approach for imbalanced learning, 2008.

With online Borderline-SMOTE, a discriminative model is not created. Instead, datapoints in the minority class are weighted according to their density, then those datapoints with the lowest density are the focus for the SMOTE synthetic example generation process.

The key idea of ADASYN algorithm is to use a density distribution as a criterion to automatically decide the number of synthetic samples that need to be generated for each minority datapoints.


In [56]:
from imblearn.over_sampling import ADASYN

y = data["target"]
X = data.drop("target", axis=1)
ordinal = OrdinalEncoder()
X_categorical = ordinal.fit_transform(X.iloc[:,0:3])
imp = SimpleImputer(strategy="mean")
X_numerical = imp.fit_transform(X.iloc[:,4:])
X_all = np.concatenate((X_categorical, X_numerical), axis=1)

# transform the dataset
oversample = ADASYN()
X_all, y = oversample.fit_resample(X_all, y)

# define model and evaluate it
model = DecisionTreeClassifier()
cv = RepeatedStratifiedKFold(n_splits=10, n_repeats=3, random_state=1)
scores = cross_val_score(model, X_all, y, scoring='roc_auc', cv=cv, n_jobs=-1)
print('Mean ROC AUC: %.3f' % mean(scores))

Mean ROC AUC: 0.943


Unlike Borderline-SMOTE, the datapoints that have the most class overlap have the most focus.
On problems where these low density datapoints might be outliers, the ADASYN approach may put too much attention on these areas of the feature space, which may result in worse model performance.

It may help to remove outliers prior to applying the oversampling procedure, and this might be a helpful heuristic to use more generally.